### PR TITLE
HSEARCH-3910 SPI for integrators to provide their own executors

### PR DESCRIPTION
**Pull request by marko-bekhta**
https://hibernate.atlassian.net/browse/HSEARCH-3910

The executor interface ended up looking like this:
```java
public interface BackendScheduledExecutor {
        // turned out we also need to have a simple `submit(..)` as `BatchingExecutor` and `LuceneParallelWorkOrchestratorImpl` 
        // are using the `BackendThreads#getWriteExecutor()` for their work.
	Future<?> submit(Runnable task);
        // the method we initially talked about
	ScheduledFuture<?> schedule(Runnable command, long delay, TimeUnit unit);
        // even though the original method returns a list of not started tasks - we never used it.
	void shutdownNow();
}
```
if the interfaces and property are ok overall, I'll add some tests to make sure the property is working ok.

### CODE REVIEW

**Comment by marko-bekhta** [comment]
Had to make the default provider bean references backend-specific. Otherwise, having multiple backends on the classpath would fail on initialization, as there'll be more than one bean with the same name (for this new provider).

File: backend/lucene/src/main/java/org/hibernate/search/backend/lucene/resources/impl/BackendThreads.java
Comment (yrodiere) [inline backend/lucene/src/main/java/org/hibernate/search/backend/lucene/resources/impl/BackendThreads.java:46]: `BeanHolder` is closable and should be closed; please see how beanholders are handled elsewhere, for example there:

https://github.com/hibernate/hibernate-search/blob/2dbdfaad4b4b75483091f7d9485ed5e1d2c77e34/backend/elasticsearch-aws/src/main/java/org/hibernate/search/backend/elasticsearch/aws/impl/ElasticsearchAwsHttpClientConfigurer.java#L93-L96

**Review by yrodiere — COMMENTED** [review]
A few comments on the implementation; I'll let @fax4ever comment on how relevant the SPI is to Infinispan :)

**Comment by fax4ever** [comment]
> A few comments on the implementation; I'll let @fax4ever comment on how relevant the SPI is to Infinispan :)

It is definitely relevant  👍

File: backend/elasticsearch/src/main/java/org/hibernate/search/backend/elasticsearch/cfg/spi/ElasticsearchBackendImplSettings.java
Comment (yrodiere) [inline backend/elasticsearch/src/main/java/org/hibernate/search/backend/elasticsearch/cfg/spi/ElasticsearchBackendImplSettings.java:42]: Shouldn't this be in SPI settings rather than impl settings? It's not just for our own benefit or for tests, it's supposed to be used by integrators. Well, we don't have any at the moment, but still... It might be useful in Quarkus, for example.

File: backend/elasticsearch/src/main/java/org/hibernate/search/backend/elasticsearch/resources/impl/ElasticsearchElasticsearchWorkExecutorProvider.java
Comment (yrodiere) [inline backend/elasticsearch/src/main/java/org/hibernate/search/backend/elasticsearch/resources/impl/ElasticsearchElasticsearchWorkExecutorProvider.java:16]: `DefaultElasticsearchWorkExecutorProvider`, maybe?

File: backend/elasticsearch/src/main/java/org/hibernate/search/backend/elasticsearch/work/spi/ElasticsearchWorkExecutorProvider.java
Comment (yrodiere) [inline backend/elasticsearch/src/main/java/org/hibernate/search/backend/elasticsearch/work/spi/ElasticsearchWorkExecutorProvider.java:14]: :+1:

File: backend/lucene/src/main/java/org/hibernate/search/backend/lucene/resources/impl/LuceneLuceneWorkExecutorProvider.java
Comment (yrodiere) [inline backend/lucene/src/main/java/org/hibernate/search/backend/lucene/resources/impl/LuceneLuceneWorkExecutorProvider.java:16]: `DefaultLuceneWorkExecutorProvider`, maybe?

**Review by yrodiere — COMMENTED** [review]
Just a few comments before you merge :) Thanks!

File: backend/elasticsearch/src/main/java/org/hibernate/search/backend/elasticsearch/cfg/spi/ElasticsearchBackendImplSettings.java
Comment (marko-bekhta) [reply backend/elasticsearch/src/main/java/org/hibernate/search/backend/elasticsearch/cfg/spi/ElasticsearchBackendImplSettings.java:42]: So if I got it correctly - just add another settings class next to this one but name it properly, right?

Comment (marko-bekhta) [reply backend/elasticsearch/src/main/java/org/hibernate/search/backend/elasticsearch/cfg/spi/ElasticsearchBackendImplSettings.java]: on it 😄

Comment (marko-bekhta) [reply backend/elasticsearch/src/main/java/org/hibernate/search/backend/elasticsearch/cfg/spi/ElasticsearchBackendImplSettings.java]: and done

File: backend/elasticsearch/src/main/java/org/hibernate/search/backend/elasticsearch/cfg/spi/ElasticsearchBackendSpiSettings.java
Comment (yrodiere) [inline backend/elasticsearch/src/main/java/org/hibernate/search/backend/elasticsearch/cfg/spi/ElasticsearchBackendSpiSettings.java:17]: Well, no :)

```suggestion
 * Configuration properties for the Elasticsearch backend that are considered SPI (and not API).
```

**Review by yrodiere — APPROVED** [review]

**Comment by yrodiere** [comment]
Failures look unrelated, merging

**Comment by fax4ever** [comment]
Thanks for the feature!

**Pull request merged**
